### PR TITLE
feat(report): Add `InstalledFiles` field to Package

### PR DESCRIPTION
**Pull request by AliDatadog**
## Description

Associate to every package the list of files that are provided by this package for apk packages only.

## Related issues
- https://github.com/aquasecurity/trivy/issues/3238

## Related PRs
- https://github.com/aquasecurity/trivy/pull/3239 This PR is a shorter version of #3239. As requested [here](https://github.com/aquasecurity/trivy/pull/3239#issuecomment-1357364465), we made it optional and split the PR per package manager.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/types/artifact.go
Comment (knqyf263) [inline pkg/fanal/types/artifact.go:101]:
```suggestion
	SystemInstalledFiles []string `json:",omitempty`
```

File: pkg/fanal/types/artifact.go
Comment (knqyf263) [inline pkg/fanal/types/artifact.go:211]: Why do we need this field?

Comment (AliDatadog) [reply pkg/fanal/types/artifact.go:211]: Good catch! It was added by mistake. It will be fixed along with the other comment in the next commit.

File: integration/testdata/alpine-310-registry.json.golden
Comment (knqyf263) [inline integration/testdata/alpine-310-registry.json.golden:1]: Seems like this file doesn't have to be updated

File: integration/testdata/conda-cyclonedx.json.golden
Comment (knqyf263) [inline integration/testdata/conda-cyclonedx.json.golden:1]: ditto

File: integration/testdata/conda-spdx.json.golden
Comment (knqyf263) [inline integration/testdata/conda-spdx.json.golden:1]: ditto

File: pkg/fanal/analyzer/pkg/apk/apk.go
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk.go:77]: Is it needed? `pkg` seems to be already reset above.

File: pkg/fanal/analyzer/pkg/apk/apk.go
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk.go:101]: Looks like an absolute path, no?

File: pkg/fanal/analyzer/pkg/apk/apk_test.go
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk_test.go:405]: `KeepSystemInstalledFiles` shouldn't always be true. We should pass the value for each test case. Otherwise, we cannot test the case where `KeepSystemInstalledFiles` is false.

**Comment by AliDatadog** [comment]
Hi @knqyf263 , thank you for the review 🙇  I addressed the comments and applied the changes.

File: pkg/fanal/analyzer/pkg/apk/apk.go
Comment (AliDatadog) [reply pkg/fanal/analyzer/pkg/apk/apk.go]: Indeed, I changed it

File: pkg/fanal/analyzer/pkg/apk/apk.go
Comment (AliDatadog) [reply pkg/fanal/analyzer/pkg/apk/apk.go]: It should be removed as well

File: docs/docs/references/configuration/cli/trivy_filesystem.md
Comment (AnaisUrlichs) [inline docs/docs/references/configuration/cli/trivy_filesystem.md:49]: do we need to further document the usage of the flag?

Comment (AliDatadog) [reply docs/docs/references/configuration/cli/trivy_filesystem.md]: I changed it to`--retain-system-installed-files     retains the files installed by each package in the analysis output when set to true.`
Do we need to add anything ?

**Comment by AliDatadog** [comment]
@knqyf263 would you have any update regarding this PR ?

File: pkg/fanal/types/artifact.go
Comment (knqyf263) [inline pkg/fanal/types/artifact.go:102]: nit: This field is located under `Package`, so we can simply call it `InstalledFiles`.
```suggestion
	InstalledFiles []string `json:",omitempty"`
```

File: pkg/fanal/analyzer/pkg/apk/apk_test.go
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk_test.go:414]: This mistake is not from your change, but we should close this file.

```suggestion
			f, err := os.Open(tt.path)
			defer f.Close()
```

File: pkg/flag/scan_flags.go
Comment (knqyf263) [inline pkg/flag/scan_flags.go:81]: Does `pkg` make more sense?

```suggestion
		Name:       "retain-pkg-installed-files",
```

File: pkg/fanal/analyzer/pkg/apk/apk.go
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk.go:100]: We should merge these two values since `SystemInstalledFiles` is now just a collection of installed files for all packages. It doesn't make sense to have them differently. But we can refactor that after merging the PR.

**Comment by AliDatadog** [comment]
Hi, any update ? @AliDatadog 🙇

**Comment by knqyf263** [comment]
After careful consideration, I thought that `--retain-pkg-installed-files` may not be necessary, as installed files are not output by default without `--list-all-pkgs`. I am sorry that you went to the trouble of implementing this. Other than that, it looks good, and I will merge it right after removing `--retain-pkg-installed-files`.

**Comment by AliDatadog** [comment]
I changed the git history to add the co-author

**Comment by knqyf263** [comment]
Tests are failing. Please let us know if you need any help.

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
>It appears that mage test:updateGolden is not updating the golden files as expected when I run the command.

The build constraint excludes the file for macOS.
https://github.com/aquasecurity/trivy/blob/cbbd1ce1f07ea2a440c08b6902958f0af6977b25/pkg/fanal/test/integration/containerd_test.go#L1

**Pull request merged**
